tests/integration/opup_test.py

```diff
@@ -0,0 +1,166 @@
+from tests.blackbox import BlackboxWrapper
+
+from algosdk.v2client.models import Account
+import algosdk
+
+
+def _dryrun(
+    bw: BlackboxWrapper,
+    sp: algosdk.future.transaction.SuggestedParams,
+    accounts: list[Account],
+) -> DryRunInspector:
+    e = PyTealDryRunExecutor(bw, pt.Mode.Application)
+    return DryRunExecutor.execute_one_dryrun(
+        tests.blackbox.algod_with_assertion(),
+        e.compile(pt.compiler.MAX_PROGRAM_VERSION),
+        [],
+        ExecutionMode.Application,
+        e.abi_argument_types(),
+        e.abi_return_type(),
+        txn_params=DryRunExecutor.transaction_params(
+            sender=graviton.models.ZERO_ADDRESS,
+            sp=sp,
+            index=DryRunExecutor.EXISTING_APP_CALL,
+            on_complete=algosdk.future.transaction.OnComplete.NoOpOC,
+        ),
+        accounts=accounts,
+    )
+
+
+_application_opcode_budget = 700
+
+
+@pytest.mark.parametrize("source", pt.OpUpFeeSource)
+@pytest.mark.parametrize("inner_txn_count", range(1, 5))
+@pytest.mark.parametrize("with_funding", [True, False])
+def test_opup_maximize_budget(
+    source: pt.OpUpFeeSource, inner_txn_count: int, with_funding: bool
+):
+    innerTxnFeeMicroAlgos = inner_txn_count * algosdk.constants.min_txn_fee
+
+    @Blackbox(input_types=[])
+    @pt.Subroutine(pt.TealType.uint64)
+    def maximize_budget():
+        return pt.Seq(
+            pt.OpUp(mode=pt.OpUpMode.OnCall).maximize_budget(
+                pt.Int(innerTxnFeeMicroAlgos), source
+            ),
+            pt.Global.opcode_budget(),
+        )
+
+    if with_funding:
+        accounts = (
+            [
+                Account(
+                    address=algosdk.logic.get_application_address(
+                        DryRunExecutor.EXISTING_APP_CALL
+                    ),
+                    status="Offline",
+                    amount=innerTxnFeeMicroAlgos,
+                    amount_without_pending_rewards=innerTxnFeeMicroAlgos,
+                )
+            ]
+            if source == pt.OpUpFeeSource.AppAccount
+            else []
+        )
+
+        sp = DryRunExecutor.SUGGESTED_PARAMS
+        sp.fee = (
+            innerTxnFeeMicroAlgos + algosdk.constants.min_txn_fee
+            if pt.OpUpFeeSource.GroupCredit
+            else sp.fee
+        )
+
+        result = _dryrun(maximize_budget, sp, accounts)
+
+        assert result.passed()
+        assert result.budget_added() == _application_opcode_budget * inner_txn_count
+    else:
+        # Withholding account and/or transaction fee funding fails the
+        # transaction.
+        sp = DryRunExecutor.SUGGESTED_PARAMS
+        sp.flat_fee = True
+        sp.fee = algosdk.constants.min_txn_fee
+        result = _dryrun(maximize_budget, DryRunExecutor.SUGGESTED_PARAMS, [])
+        assert not result.passed()
+
+
+@pytest.mark.parametrize("source", [f for f in pt.OpUpFeeSource])
+@pytest.mark.parametrize("budget_added", list(range(1_000, 20_000, 2_500)))
```

Comment:
similar suggestions to previousing regarding iterators vs. lists

Reply:
@tzaffi Can you see if https://github.com/algorand/pyteal/pull/607/commits/5bfe54bde38afbbffccc36072d2a7cf87212811d addresses the feedback?